### PR TITLE
Add new admin panel v2.0

### PR DESCRIPTION
**Pull request by IvanBuljovcic**
Adds features to the admin panel to update partymap template

### CODE REVIEW

**Comment by mike-lvov** [comment]
A lot of things happening here. Seems like it's working, but it is too hard to get a complete understanding of what is going on here because of the PR's size. I have quickly skimmed through this and noted a few things, but tbh, I think we will have to refactor it later. Having thought about that, we could just merge it in for now as it is bringing value for the users and just rewrite it slowly later on.

@thecadams @0xdevalias definitely need your thoughts on this one

File: src/components/atoms/Button/Button.tsx
Comment (IvanBuljovcic) [reply src/components/atoms/Button/Button.tsx:52]: Brain fart moment... will be fixed

File: src/components/atoms/ImageInput/ImageInput.tsx
Comment (IvanBuljovcic) [reply src/components/atoms/ImageInput/ImageInput.tsx:45]: Another brain fart, this was fixed and then lost in the great commit that made me close the previous 3 PRs..

the imgUrl is being set as the default state in the latest fix (to be committed)

File: src/pages/Admin/Room/Card/RoomCard.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Room/Card/RoomCard.tsx:23]: `useMemo`

File: src/pages/Admin/Room/Modal/Item/Item.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Room/Modal/Item/Item.tsx:29]: useCallback

File: src/pages/Admin/Room/Modal/Item/Item.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Room/Modal/Item/Item.tsx:74]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:101]: useCallback

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:104]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:110]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:115]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:182]: Function + useCallback.

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:208]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:209]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:215]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:218]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:225]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:229]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:238]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:245]: .

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:247]: .

File: src/pages/Admin/Venue/QuestionInput.tsx
Comment (denisdimitrov) [inline src/pages/Admin/Venue/QuestionInput.tsx:12]: Remove `any` use type.

File: src/pages/Admin/Venue/VenueWizard/Create/VenueWizardCreate.types.ts
Comment (denisdimitrov) [inline src/pages/Admin/Venue/VenueWizard/Create/VenueWizardCreate.types.ts:6]: Remove `any` Dispatch type should exist in react or react-redux

File: src/pages/Admin/Venue/VenueWizard/Edit/VenueWizardEdit.types.ts
Comment (denisdimitrov) [inline src/pages/Admin/Venue/VenueWizard/Edit/VenueWizardEdit.types.ts:7]: .

File: src/pages/Admin/Venue/VenueWizard/redux/actions.ts
Comment (denisdimitrov) [inline src/pages/Admin/Venue/VenueWizard/redux/actions.ts:9]: .

File: src/pages/Admin/Venue/VenueWizard/redux/actions.ts
Comment (denisdimitrov) [inline src/pages/Admin/Venue/VenueWizard/redux/actions.ts:23]: Remove all of the disables and add types.

File: src/pages/Admin/Details/Form/DetailsForm.tsx
Comment (0xdevalias) [reply src/pages/Admin/Details/Form/DetailsForm.tsx:236]: Without looking deeper into the file/specifics of each, my preference would be to use smaller components if needed/similar.

File: src/pages/Admin/Details/Preview/DetailsPreview.tsx
Comment (0xdevalias) [reply src/pages/Admin/Details/Preview/DetailsPreview.tsx:32]: Agree, preference for components over render functions

File: src/pages/Admin/Details/ValidationSchema.ts
Comment (0xdevalias) [reply src/pages/Admin/Details/ValidationSchema.ts:44]: I wouldn't manually add another string to settings. Instead I would calculate it from the actual value being used in the validation: `MAX_IMAGE_FILE_SIZE_BYTES`

Something like: `MAX_IMAGE_FILE_SIZE_BYTES / 1024 / 1024` or similar

File: src/pages/Admin/MapPreview/MapPreview.tsx
Comment (0xdevalias) [reply src/pages/Admin/MapPreview/MapPreview.tsx:138]: Guessing you mean the `@ts-ignore` should be removed from here + the associated things it was ignoring fixed?

File: src/pages/Admin/Room/Card/RoomCard.tsx
Comment (0xdevalias) [reply src/pages/Admin/Room/Card/RoomCard.tsx:23]: Also use selector function or 'wrapping hook' instead of the inline function passed to `useSelector` as per https://github.com/sparkletown/internal-sparkle-issues/issues/115

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (0xdevalias) [reply src/pages/Admin/Venue/Details/VenueDetails.tsx:115]: Also probably use semantic names for these functions. I haven't deeply looked at what is being done here, but `handleEditRoomSave` could probably just be `saveRoom` or similar.

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (0xdevalias) [reply src/pages/Admin/Venue/Details/VenueDetails.tsx:182]: In this instance, since there are no parameters that need to be passed to `toggleRoomModal`, you can actually just do:

```suggestion
              <Button onClick={toggleRoomModal} gradient>
```

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (0xdevalias) [reply src/pages/Admin/Venue/Details/VenueDetails.tsx:209]: Probably call this `closeModal` or similar

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (0xdevalias) [reply src/pages/Admin/Venue/Details/VenueDetails.tsx:215]: Maybe call this `cancelRoomEditing`/`cancelEditingRoom` or something like that?

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (0xdevalias) [reply src/pages/Admin/Venue/Details/VenueDetails.tsx:218]: Probably call this `showDeleteModal`

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (0xdevalias) [reply src/pages/Admin/Venue/Details/VenueDetails.tsx:245]: `hideEventModal`

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (0xdevalias) [reply src/pages/Admin/Venue/Details/VenueDetails.tsx:247]: Probably `noopFunc` or make these params optional.

I would also consider whether there is a better 'higher level' function we could be passing down instead of the raw `setEditedEvent`, `setShowDeleteEventModal`, etc.

File: src/pages/Admin/Venue/VenueWizard/Create/VenueWizardCreate.types.ts
Comment (0xdevalias) [reply src/pages/Admin/Venue/VenueWizard/Create/VenueWizardCreate.types.ts:6]: Depending on what the 'meaning' of this `dispatch` is, we may just not need it. If it is just the normal redux `dispatch`, then we should probably just use the `useDispatch` hook in the component where we need it, rather that pass it around as props.

File: src/pages/Admin/Venue/VenueWizard/Edit/VenueWizardEdit.types.ts
Comment (0xdevalias) [reply src/pages/Admin/Venue/VenueWizard/Edit/VenueWizardEdit.types.ts:7]: https://github.com/sparkletown/sparkle/pull/1056#discussion_r553097439

File: src/pages/Admin/Venue/VenueWizard/redux/actions.ts
Comment (0xdevalias) [reply src/pages/Admin/Venue/VenueWizard/redux/actions.ts:9]: https://github.com/sparkletown/sparkle/pull/1056#discussion_r553097439

File: src/pages/Admin/Details/Form/DetailsForm.tsx
Comment (mike-lvov) [inline src/pages/Admin/Details/Form/DetailsForm.tsx:49]: ?

File: src/pages/Admin/MapPreview/MapPreview.tsx
Comment (mike-lvov) [inline src/pages/Admin/MapPreview/MapPreview.tsx:34]: Is it possible to remove `!` to ensure type safety across the app?

File: src/pages/Admin/MapPreview/MapPreview.tsx
Comment (mike-lvov) [inline src/pages/Admin/MapPreview/MapPreview.tsx:52]: This looks a bit hacky. is it a WIP implementation or is it going to production?

File: src/pages/Admin/MapPreview/MapPreview.tsx
Comment (mike-lvov) [inline src/pages/Admin/MapPreview/MapPreview.tsx:59]: ?

File: src/pages/Admin/MapPreview/MapPreview.tsx
Comment (mike-lvov) [inline src/pages/Admin/MapPreview/MapPreview.tsx:70]: This looks a bit hacky. is it a WIP implementation or is it going to production?

File: src/pages/Admin/MapPreview/MapPreview.tsx
Comment (mike-lvov) [inline src/pages/Admin/MapPreview/MapPreview.tsx:127]: higher in the code you have `mapRooms!`, which is bug-prone. Has something changed here that you use `?`?

File: src/pages/Admin/Venue/Details/VenueDetails.tsx
Comment (mike-lvov) [inline src/pages/Admin/Venue/Details/VenueDetails.tsx:103]: I would move this to a separate service\util function\class to abstract from direct `window` usage. But it is possible that this can be done in a separate PR not to block this PR's features

**Review by mike-lvov — CHANGES_REQUESTED** [review]
I have only reviewed the recent changes. Maybe a few fixes should occur to mark this PR as approved.

File: src/pages/Admin/MapPreview/MapPreview.tsx
Comment (0xdevalias) [reply src/pages/Admin/MapPreview/MapPreview.tsx:34]: Agreed. This is something that comes up relatively often. It would be good to see if there is a way to eslint ban it.

Comment (0xdevalias) [reply src/pages/Admin/MapPreview/MapPreview.tsx:34]: Backlogged this general concept in https://github.com/sparkletown/internal-sparkle-issues/issues/149

File: src/pages/Admin/MapPreview/MapPreview.tsx
Comment (0xdevalias) [reply src/pages/Admin/MapPreview/MapPreview.tsx:52]: What's the purpose/intent of this bit of code?

File: src/pages/Admin/MapPreview/MapPreview.types.ts
Comment (0xdevalias) [reply src/pages/Admin/MapPreview/MapPreview.types.ts:8]: Or `withEditing` since it's used in props here is another way we have isued this.

File: src/pages/Admin/MapPreview/MapPreview.tsx
Comment (denisdimitrov) [reply src/pages/Admin/MapPreview/MapPreview.tsx:52]: @0xdevalias The purpose is to create a new array of rooms but with the new updated positions and set them in the state. That way I can use the updated state whilst the locations are being updated otherwise, after the first one is updated the props change and the rooms go back to their old positions. It's a very unpleasant UI flicker.

Comment (0xdevalias) [reply src/pages/Admin/MapPreview/MapPreview.tsx:52]: It looks like the rooms (`roomRef`) is an `object`, but the way you are getting the values and mapping through it makes it feel like it should be an `array`?

I don't deeply have my 🧠  in this code to be able to suggest if there is a better way to handle this. Just feels like it's a little convoluted at the moment maybe? Happy to go with your / @mike-lvov 's judgement on it though.

File: src/components/atoms/Button/Button.tsx
Comment (mike-lvov) [inline src/components/atoms/Button/Button.tsx:26]: [NOT BLOCKING]: Would be nice to have loaders separated into a component

File: src/components/atoms/Button/Button.tsx
Comment (mike-lvov) [reply src/components/atoms/Button/Button.tsx:51]: [NB]: Should be fixed, I guess

File: src/components/atoms/Button/Button.tsx
Comment (mike-lvov) [inline src/components/atoms/Button/Button.tsx:57]: [NB] I think we should stick to one way of setting defultProps

File: src/components/atoms/ImageInput/ImageInput.types.ts
Comment (mike-lvov) [reply src/components/atoms/ImageInput/ImageInput.types.ts:10]: [NB] True that

File: src/hooks/useAdminVenues.ts
Comment (mike-lvov) [reply src/hooks/useAdminVenues.ts:14]: NB ^

**Review by mike-lvov — APPROVED** [review]
A few non-blocking comments there to address after it is merged. Good job! 
I hope, we will not need to have the PRs of this size. It's mentally hard to review it TBH

File: src/components/atoms/Button/Button.tsx
Comment (0xdevalias) [reply src/components/atoms/Button/Button.tsx:57]: Agreed that we should only have one way. From memory there were a few advantages that can be gained by doing it this way.. but i'm not sure we really care about them. Unless there is good evidence as to why we should do it differently, my personal preference is to use standard ES6+ syntax

**Comment by 0xdevalias** [comment]
@denisdimitrov probably worth merging `staging` into this now that https://github.com/sparkletown/sparkle/pull/1070 has been merged, then we could do clickthroughs/refinements/etc with @sparkleed before merging into `staging` if desired.

File: src/hooks/useAdminVenues.ts
Comment (0xdevalias) [inline src/hooks/useAdminVenues.ts:7]: I expect the `userUid || ""` pattern here was originally to avoid cases where `userUid` is `undefined`, but the way it is defaulted here is probably a bug/potential security issue if there was ever an empty string (`""`) in the owners array of a venue.

We should fix this so that it aligns with our current patterns, selectively connecting the query only when all of it's 'required bits' exist, and not running it (eg. making the whole query `undefined`) if not.

Comment (0xdevalias) [reply src/hooks/useAdminVenues.ts:7]: backlogged in https://github.com/sparkletown/internal-sparkle-issues/issues/185

**Pull request merged**
